crates/sui-json-rpc/src/routing_layer.rs

```diff
@@ -0,0 +1,138 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::metrics::is_json;
+use hyper::body::Bytes;
+use hyper::{body, http, Body, Request, Response};
+use jsonrpsee::core::__reexports::serde_json;
+use jsonrpsee::core::server::helpers::MethodResponse;
+use jsonrpsee::types::error::ErrorCode;
+use jsonrpsee::types::{ErrorObject, Request as RpcRequest};
+use std::collections::{HashMap, HashSet};
+use std::error::Error;
+use std::future::Future;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use sui_open_rpc::MethodRouting;
+use tower::{Layer, Service};
+
+#[derive(Debug, Clone)]
+pub struct RoutingLayer {
+    routes: HashMap<String, MethodRouting>,
+    disable_routing: bool,
+}
+
+impl RoutingLayer {
+    pub fn new(routes: HashMap<String, MethodRouting>, disable_routing: bool) -> Self {
+        Self {
+            routes,
+            disable_routing,
+        }
+    }
+}
+
+impl<S> Layer<S> for RoutingLayer {
+    type Service = RpcRoutingService<S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        RpcRoutingService::new(inner, self.routes.clone(), self.disable_routing)
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct RpcRoutingService<S> {
+    inner: S,
+    routes: HashMap<String, MethodRouting>,
+    route_to_methods: HashSet<String>,
+    disable_routing: bool,
+}
+
+impl<S> RpcRoutingService<S> {
+    pub fn new(inner: S, routes: HashMap<String, MethodRouting>, disable_routing: bool) -> Self {
+        let route_to_methods = routes.iter().map(|(_, v)| v.route_to.clone()).collect();
+        Self {
+            inner,
+            routes,
+            route_to_methods,
+            disable_routing,
+        }
+    }
+}
+
+impl<S> Service<Request<Body>> for RpcRoutingService<S>
+where
+    S: Service<Request<Body>, Response = Response<Body>> + Clone + Send + 'static,
+    S::Response: 'static,
+    S::Error: Into<Box<dyn Error + Send + Sync>> + 'static,
+    S::Future: Send + 'static,
+{
+    type Response = S::Response;
+    type Error = Box<dyn Error + Send + Sync + 'static>;
+    type Future =
+        Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send + 'static>>;
+
+    #[inline]
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx).map_err(Into::into)
+    }
+
+    fn call(&mut self, req: Request<Body>) -> Self::Future {
+        let clone = self.inner.clone();
+        let routes = self.routes.clone();
+        let route_to_methods = self.route_to_methods.clone();
+        let disable_routing = self.disable_routing;
+        // take the service that was ready
+        // https://docs.rs/tower/latest/tower/trait.Service.html#be-careful-when-cloning-inner-services
+        let mut inner = std::mem::replace(&mut self.inner, clone);
+        let res_fut = async move {
+            // Get version from header.
+            let version = req
+                .headers()
+                .get("client_api_version")
```

Comment:
FYI made some renaming for this in https://github.com/MystenLabs/sui/pull/8394. Before this PR, there were also some inconsistencies in the header names(`client-api-version` instead of `client_api_version`) between TS and Rust, which might be the reason why e2e tests are failing

Reply:
thanks for changing this! I will wait for your PR before merging this one.